### PR TITLE
[Core] Support custom image on Azure

### LINKED ISSUE #2910: Cannot choose Custom V-net and custom images in Azure
I cannot choose Custom V-net and custom images in Azure, is there a specific format I have to use, or is the feature not available for Azure?

(sky) skypilot@kenadyinampudi:$ sky launch skypilot_azure.yaml
ValueError: Invalid config YAML (/home/skypilot/.sky/config.yaml): The following fields are invalid:
Found unsupported field 'azure'.
(sky) skypilot@kenadyinampudi:$ sky launch skypilot_azure.yaml
Task from YAML spec: skypilot_azure.yaml
ValueError: image_id is only supported for AWS/GCP/IBM/OCI, please explicitly specify the cloud.
(sky) skypilot@kenadyinampudi:~$

### PR DESCRIPTION
**Pull request by cblmemo**
This PR adds support for custom images on Azure. Partially fixes #2910.

Related: skypilot-org/skypilot-catalog#66

Tested (run the relevant ones):

- [x] Code formatting: `bash format.sh`
- [x] Any manual or new tests for this PR (please specify below)
     - [x] `sky launch --cloud azure --image-id microsoft-dsvm:ubuntu-2004:2004:21.11.04` and check the image on the cloud console
- [ ] All smoke tests: `pytest tests/test_smoke.py` 
- [x] Relevant individual smoke tests: `pytest tests/test_smoke.py::test_fill_in_the_name` 
    - [x] pytest tests/test_smoke.py::test_azure_images
- [ ] Backward compatibility tests: `bash tests/backward_comaptibility_tests.sh`

### CODE REVIEW

File: sky/clouds/azure.py [hunk 1]
```diff
@@ -137,6 +134,43 @@ def get_egress_cost(self, num_gigabytes: float):
...
 
+    @classmethod
+    def get_image_size(cls, image_id: str, region: Optional[str]) -> float:
+        if region is None:
+            # The region used here is only for where to send the query,
+            # not the image location. Azure's image is globally available.
+            region = 'eastus'
+        if image_id.startswith('skypilot:'):
+            image_id = service_catalog.get_image_id_from_tag(image_id,
+                                                             clouds='azure')
```
Comment (Michaelvll) [inline sky/clouds/azure.py:145]: For the tagged images, let's shortcircuit the size check by directly setting the image size to the number we know. This is to optimize the speed for creating `Resources` object in most cases.

File: tests/test_optimizer_dryruns.py [hunk 2]
```diff
@@ -468,8 +468,8 @@ def test_invalid_image(monkeypatch):
     assert 'Cloud must be specified' in str(e.value)
```
Comment (Michaelvll) [inline tests/test_optimizer_dryruns.py:468]: Should we add a smoke test for the image_id for Azure as well?

**Review by Michaelvll — COMMENTED** [review]
Thanks for adding the support for Azure custom image @cblmemo! This is awesome. Left several comments.

File: docs/source/reference/yaml-spec.rst [hunk 3]
```diff
@@ -194,6 +194,10 @@ Available fields:
       # Or machine image: https://cloud.google.com/compute/docs/machine-images
       # image_id: projects/my-project/global/machineImages/my-machine-image
       #
+      # Azure
+      # To find Azure images: https://docs.microsoft.com/en-us/azure/virtual-machines/linux/cli-ps-findimage
+      # image_id: microsoft-dsvm:ubuntu-2004:2004:21.11.04
```
[the remark this answers is not in the document]
Comment (cblmemo) [reply docs/source/reference/yaml-spec.rst:199]: this is the standard format on Azure (look at the fifth column):

```bash
az vm image list --publisher microsoft-dsvm --all --output table | grep ubuntu-hpc
x64             ubuntu-hpc                    microsoft-dsvm  1804                      microsoft-dsvm:ubuntu-hpc:1804:18.04.2021051701                          18.04.2021051701
x64             ubuntu-hpc                    microsoft-dsvm  1804                      microsoft-dsvm:ubuntu-hpc:1804:18.04.2021110101                          18.04.2021110101
x64             ubuntu-hpc                    microsoft-dsvm  1804                      microsoft-dsvm:ubuntu-hpc:1804:18.04.2021120101                          18.04.2021120101
x64             ubuntu-hpc                    microsoft-dsvm  1804                      microsoft-dsvm:ubuntu-hpc:1804:18.04.2022061601                          18.04.2022061601
x64             ubuntu-hpc                    microsoft-dsvm  1804                      microsoft-dsvm:ubuntu-hpc:1804:18.04.2022121201                          18.04.2022121201
x64             ubuntu-hpc                    microsoft-dsvm  1804                      microsoft-dsvm:ubuntu-hpc:1804:18.04.2023031501                          18.04.2023031501
x64             ubuntu-hpc                    microsoft-dsvm  1804                      microsoft-dsvm:ubuntu-hpc:1804:18.04.2023063001                          18.04.2023063001
x64             ubuntu-hpc                    microsoft-dsvm  2004                      microsoft-dsvm:ubuntu-hpc:2004:20.04.2021051401                          20.04.2021051401
x64             ubuntu-hpc                    microsoft-dsvm  2004                      microsoft-dsvm:ubuntu-hpc:2004:20.04.2021110101                          20.04.2021110101
x64             ubuntu-hpc                    microsoft-dsvm  2004                      microsoft-dsvm:ubuntu-hpc:2004:20.04.2021120101                          20.04.2021120101
x64             ubuntu-hpc                    microsoft-dsvm  2004                      microsoft-dsvm:ubuntu-hpc:2004:20.04.2022121201                          20.04.2022121201
x64             ubuntu-hpc                    microsoft-dsvm  2004                      microsoft-dsvm:ubuntu-hpc:2004:20.04.2023011101                          20.04.2023011101
x64             ubuntu-hpc                    microsoft-dsvm  2004                      microsoft-dsvm:ubuntu-hpc:2004:20.04.2023031501                          20.04.2023031501
x64             ubuntu-hpc                    microsoft-dsvm  2004                      microsoft-dsvm:ubuntu-hpc:2004:20.04.2023063001                          20.04.2023063001
x64             ubuntu-hpc                    microsoft-dsvm  2004                      microsoft-dsvm:ubuntu-hpc:2004:20.04.2023080201                          20.04.2023080201
x64             ubuntu-hpc                    microsoft-dsvm  2004                      microsoft-dsvm:ubuntu-hpc:2004:20.04.2023111801                          20.04.2023111801
x64             ubuntu-hpc                    microsoft-dsvm  2004-preview-ndv5         microsoft-dsvm:ubuntu-hpc:2004-preview-ndv5:20.04.2023041401             20.04.2023041401
x64             ubuntu-hpc                    microsoft-dsvm  2004-preview-ndv5         microsoft-dsvm:ubuntu-hpc:2004-preview-ndv5:20.04.2023041701             20.04.2023041701
x64             ubuntu-hpc                    microsoft-dsvm  2004-preview-ndv5         microsoft-dsvm:ubuntu-hpc:2004-preview-ndv5:20.04.2023042401             20.04.2023042401
x64             ubuntu-hpc                    microsoft-dsvm  2004-preview-ndv5         microsoft-dsvm:ubuntu-hpc:2004-preview-ndv5:20.04.2023051001             20.04.2023051001
x64             ubuntu-hpc                    microsoft-dsvm  2004-preview-ndv5         microsoft-dsvm:ubuntu-hpc:2004-preview-ndv5:20.04.2023063001             20.04.2023063001
x64             ubuntu-hpc                    microsoft-dsvm  2004-preview-ndv5         microsoft-dsvm:ubuntu-hpc:2004-preview-ndv5:20.04.2023080201             20.04.2023080201
x64             ubuntu-hpc                    microsoft-dsvm  2204                      microsoft-dsvm:ubuntu-hpc:2204:22.04.2023041401                          22.04.2023041401
x64             ubuntu-hpc                    microsoft-dsvm  2204                      microsoft-dsvm:ubuntu-hpc:2204:22.04.2023041701                          22.04.2023041701
x64             ubuntu-hpc                    microsoft-dsvm  2204                      microsoft-dsvm:ubuntu-hpc:2204:22.04.2023042401                          22.04.2023042401
x64             ubuntu-hpc                    microsoft-dsvm  2204                      microsoft-dsvm:ubuntu-hpc:2204:22.04.2023051001                          22.04.2023051001
x64             ubuntu-hpc                    microsoft-dsvm  2204                      microsoft-dsvm:ubuntu-hpc:2204:22.04.2023061401                          22.04.2023061401
x64             ubuntu-hpc                    microsoft-dsvm  2204                      microsoft-dsvm:ubuntu-hpc:2204:22.04.2023063001                          22.04.2023063001
x64             ubuntu-hpc                    microsoft-dsvm  2204                      microsoft-dsvm:ubuntu-hpc:2204:22.04.2023080201                          22.04.2023080201
x64             ubuntu-hpc                    microsoft-dsvm  2204                      microsoft-dsvm:ubuntu-hpc:2204:22.04.2023111801                          22.04.2023111801
x64             ubuntu-hpc                    microsoft-dsvm  2204-preview-ndv5         microsoft-dsvm:ubuntu-hpc:2204-preview-ndv5:22.04.2023041401             22.04.2023041401
x64             ubuntu-hpc                    microsoft-dsvm  2204-preview-ndv5         microsoft-dsvm:ubuntu-hpc:2204-preview-ndv5:22.04.2023041701             22.04.2023041701
x64             ubuntu-hpc                    microsoft-dsvm  2204-preview-ndv5         microsoft-dsvm:ubuntu-hpc:2204-preview-ndv5:22.04.2023042401             22.04.2023042401
x64             ubuntu-hpc                    microsoft-dsvm  2204-preview-ndv5         microsoft-dsvm:ubuntu-hpc:2204-preview-ndv5:22.04.2023051001             22.04.2023051001
x64             ubuntu-hpc                    microsoft-dsvm  2204-preview-ndv5         microsoft-dsvm:ubuntu-hpc:2204-preview-ndv5:22.04.2023061401             22.04.2023061401
x64             ubuntu-hpc                    microsoft-dsvm  2204-preview-ndv5         microsoft-dsvm:ubuntu-hpc:2204-preview-ndv5:22.04.2023070601             22.04.2023070601
x64             ubuntu-hpc                    microsoft-dsvm  2204-preview-ndv5         microsoft-dsvm:ubuntu-hpc:2204-preview-ndv5:22.04.2023080201             22.04.2023080201
```

File: sky/clouds/azure.py (hunk 1 above)
Comment (cblmemo) [reply sky/clouds/azure.py:145]: Good point! Added.

File: docs/source/reference/yaml-spec.rst (hunk 3 above)
[the remark this answers is not in the document]
Comment (Michaelvll) [reply docs/source/reference/yaml-spec.rst:199]: Good point! I `:` sounds good to me then.

File: tests/test_optimizer_dryruns.py (hunk 2 above)
Comment (cblmemo) [reply tests/test_optimizer_dryruns.py:468]: Added and passed. Thanks!

File: sky/adaptors/azure.py [hunk 4]
```diff
@@ -46,6 +46,13 @@ def http_error_exception():
     return exceptions.HttpResponseError
 
 
+@import_package
+def resources_not_found_exception():
+    """ResourceNotFoundError exception."""
+    from azure.core import exceptions
+    return exceptions.ResourceNotFoundError
```
Comment (Michaelvll) [inline sky/adaptors/azure.py:53]: Should we combine this and the `http_error_exception()` above into
```
@import_package
def exceptions():
    """ResourceNotFoundError exception."""
    from azure.core import exceptions
    return exceptions
```

and in all the callers, we use `azure.exceptions().specific_exception`?

File: sky/clouds/azure.py [hunk 5]
```diff
@@ -137,6 +137,51 @@ def get_egress_cost(self, num_gigabytes: float):
...
+        if len(image_id_splitted) != 4:
+            with ux_utils.print_exception_no_traceback():
+                raise ValueError(f'Invalid image id: {image_id}. Expected '
+                                 'format: <publisher>:<offer>:<sku>:<version>')
+        publisher, offer, sku, version = image_id_splitted
+        if is_skypilot_image_tag:
+            if offer == 'ubuntu-hpc':
+                return _DEFAULT_AZURE_UBUNTU_HPC_IMAGE_GB
+            else:
+                assert offer == 'ubuntu-2004', offer
```
Comment (Michaelvll) [inline sky/clouds/azure.py:161]: Let's avoid the assertion here, so that customized image (user changes images.csv manually) will not cause an error.

**Review by Michaelvll — APPROVED** [review]
Thanks for adding the support for Azure custom image @cblmemo! LGTM. Please feel free to merge it once the tests pass.

File: sky/adaptors/azure.py (hunk 4 above)
Comment (cblmemo) [reply sky/adaptors/azure.py:53]: SG! Changed.

File: sky/clouds/azure.py (hunk 5 above)
Comment (cblmemo) [reply sky/clouds/azure.py:161]: Good point! Removed.

**Pull request merged**